closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -901,6 +902,23 @@ JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_DigestCreateCon
     context->ctx = ctx;
     context->digestAlg = digestAlg;
 
+    // Create a second initalized openssl digest context. This is being done for performance reasons since
+    // creating and or re-initializing digest contexts later during processing is found to be expensive.
+    // This second context, context->cachedInitalizedDigestContext, will be copied over the working context,
+    // context->ctx, using the EVP_MD_CTX_copy_ex API whenever we wish to re-initalize this cipher. This occurs
+    // during an explicit reset of the cipher or whenever a final digest is computed.
+    if (NULL == (context->cachedInitalizedDigestContext = (*OSSL_MD_CTX_new)())) {
+        printErrors();
+        Java_jdk_crypto_jniprovider_NativeCrypto_DigestDestroyContext(env, thisObj, (jlong)(intptr_t)context);
+        return -1;
+    }
+
+    if (1 != (*OSSL_DigestInit_ex)(context->cachedInitalizedDigestContext, digestAlg, NULL)) {
+        printErrors();
+        Java_jdk_crypto_jniprovider_NativeCrypto_DigestDestroyContext(env, thisObj, (jlong)(intptr_t)context);
+        return -1;
```

Comment:
Suggest adding a label to the cleanup code (i.e. lines 911-913) to be used here and in the failure cases below (lines 925 and 930).

Reply:
Added a label for cleanup called by these lines and the former lines 925-930.